tests/ut/attention_kernels_test.cpp

```diff
@@ -0,0 +1,231 @@
+                sum += std::exp(fscore[i * tokenSize + j] - maxVal);
+            }
+            float rsum = 1.0f / sum;
+            for (int j = 0; j < size; j++) {
+                score[i * tokenSize + j] = (bfloat16_t)(std::exp(fscore[i * tokenSize + j] - maxVal) * rsum);
+            }
+            for (int j = size; j < tokenSize; ++j) {
+                score[i * tokenSize + j] = 0.0f;
+            }
+        }
+
+        // Compute matmul between result and 'value'
+        auto out = output + h * headSize;
+        mmRef(score, v, out, tokenSize, headSize, tokenSize, tokenSize, kvStride, oStride, false);
+
+        free(score);
+        free(fscore);
+    }
+}
+
+// Reference implementation of self-attention
+static void selfAttentionRef(bfloat16_t *output, bfloat16_t *query, bfloat16_t *key, bfloat16_t *value, int qHeadNum,
+        int kvHeadNum, int headSize, int oStride, int qStride, int kvStride, int batchSize, const int *tokenSizes,
+        const float scale) {
+
+    int rowOffsets[batchSize] = {0};
+    for (int i = 1; i < batchSize; i++) {
+        rowOffsets[i] = rowOffsets[i - 1] + tokenSizes[i - 1];
+    }
+
+    for (int b = 0; b < batchSize; b++) {
+        auto q = query + rowOffsets[b] * qStride;
+        auto k = key + rowOffsets[b] * kvStride;
+        auto v = value + rowOffsets[b] * kvStride;
+        selfAttentionRef(output + rowOffsets[b] * oStride, q, k, v, qHeadNum, kvHeadNum, headSize, oStride, qStride,
+                kvStride, tokenSizes[b], scale);
+    }
+}
+
+void testSelfAttention(
+        int headSize, int qHeadNum, int kvHeadNum, int *tokenSizes, int batchSize, bool bSeparateCopy = true) {
+    const int qkvStride = headSize * (qHeadNum + 2 * kvHeadNum);
+    const int hiddenSize = headSize * qHeadNum;
+    const float scale = 1.0f / std::sqrt(headSize);
+    int threadNum = -1;
+
+#pragma omp parallel
+    {
+        int tid = omp_get_thread_num();
+        if (tid == 0) { threadNum = omp_get_num_threads(); }
+    }
+
+    int totalTokens = 0;
+    int maxTokens = 0;
+    for (int i = 0; i < batchSize; i++) {
+        totalTokens += tokenSizes[i];
+        maxTokens = std::max(maxTokens, tokenSizes[i]);
+    }
+
+    // Buffers
+    bfloat16_t *ourOutput = (bfloat16_t *)xft::alloc(totalTokens * hiddenSize * sizeof(bfloat16_t));
+    bfloat16_t *refOutput = (bfloat16_t *)xft::alloc(totalTokens * hiddenSize * sizeof(bfloat16_t));
+    bfloat16_t *kvCache
+            = (bfloat16_t *)xft::alloc(2 * batchSize * maxTokens * (kvHeadNum * headSize) * sizeof(bfloat16_t));
+    bfloat16_t *qkv = (bfloat16_t *)xft::alloc(totalTokens * qkvStride * sizeof(bfloat16_t));
+
+    bfloat16_t *kCache = kvCache;
+    bfloat16_t *vCache = kvCache + batchSize * maxTokens * (kvHeadNum * headSize);
+
+    bfloat16_t *query = qkv;
+    bfloat16_t *key = qkv + headSize * qHeadNum;
+    bfloat16_t *value = qkv + headSize * (qHeadNum + kvHeadNum);
+
+    for (int i = 0; i < totalTokens * qkvStride; i++) {
```

Comment:
#pragma omp parallel for

Reply:
when use rand(), openmp will make it slower. Here is an article:
https://mahesh-hegde.github.io/posts/omp_rand_critical_section/